File: java/src/main/java/cn/az/code/year2021/mar/UndergroundSystem.java
package cn.az.code.year2021.mar;

import javafx.util.Pair;

import java.util.HashMap;
import java.util.Map;

/**
 * @author az
 * @since 03/20/21 21:33
 */
public class UndergroundSystem {

    // Route - {TotalTime, Count}

    Map<String, Pair<Integer, Integer>> checkoutMap = new HashMap<>();

    // Uid - {StationName, Time}

    Map<Integer, Pair<String, Integer>> checkInMap = new HashMap<>();

    public UndergroundSystem() {
    }

    public void checkIn(int id, String stationName, int t) {
        checkInMap.put(id, new Pair<>(stationName, t));
    }

    public void checkOut(int id, String stationName, int t) {
        Pair<String, Integer> checkIn = checkInMap.get(id);
        String route = checkIn.getKey() + "_" + stationName;
        int totalTime = t - checkIn.getValue();
        Pair<Integer, Integer> checkout = checkoutMap.getOrDefault(route, new Pair<>(0, 0));
        checkoutMap.put(route, new Pair<>(checkout.getKey() + totalTime, checkout.getValue() + 1));
    }

    public double getAverageTime(String startStation, String endStation) {
        String route = startStation + "_" + endStation;
        Pair<Integer, Integer> checkout = checkoutMap.get(route);
        return (double) checkout.getKey() / checkout.getValue();
    }
}

File: src/design/underground_system.cc
#include <string>
#include <unordered_map>

using namespace std;

class UndergroundSystem
{
public:
    UndergroundSystem()
    {
    }

    void checkIn(int id, string stationName, int t)
    {
        check_in_map_[id] = make_pair(stationName, t);
    }

    void checkOut(int id, string stationName, int t)
    {
        auto check_in_ = check_in_map_[id];
        int total_time = t - check_in_.second;
        auto key = check_in_.first + "_" + stationName;
        auto check_out_ = check_out_map_.find(key);
        if (check_out_ != check_out_map_.end())
        {
            check_out_->second.first += total_time;
            check_out_->second.second++;
        }
        else
        {
            check_out_map_[key] = make_pair(total_time, 1);
        }
    }

    double getAverageTime(string startStation, string endStation)
    {
        auto key = startStation + "_" + endStation;
        auto check_out_ = check_out_map_.find(key);
        if (check_out_ != check_out_map_.end())
        {
            return (double)check_out_->second.first / check_out_->second.second;
        }
        return 0.0;
    }

private:
    // id -> (stationName, t)
    unordered_map<int, pair<string, int>> check_in_map_;
    // (startStation + endStation) -> (total_time, count)
    unordered_map<string, pair<int, int>> check_out_map_;
};
